### DISCUSSION TITLE
OhMyZsh Plugin error "can only be called from completion function"

### DISCUSSION THREAD

**Discussion by hoyacom** [marlonrichert/zsh-snap#55] [Q&A] [answered]
This is My part of .zshrc 

```bash
  source ~/.znap/zsh-snap/znap.zsh

  #znap prompt romkatv/powerlevel10k
  znap source ohmyzsh/ohmyzsh
  znap source ohmyzsh/ohmyzsh lib/completion
  znap source ohmyzsh/ohmyzsh plugins/git
  znap source ohmyzsh/ohmyzsh plugins/command-not-found
  znap source ohmyzsh/ohmyzsh plugins/docker
  znap source ohmyzsh/ohmyzsh plugins/fasd
  znap source ohmyzsh/ohmyzsh plugins/fzf

  znap source ohmyzsh/ohmyzsh plugins/fd
  znap source ohmyzsh/ohmyzsh plugins/ripgrep

  znap source ohmyzsh/ohmyzsh plugins/osx
  znap source ohmyzsh/ohmyzsh plugins/tmux
  znap source ohmyzsh/ohmyzsh plugins/tmuxinator
  znap source ohmyzsh/ohmyzsh plugins/thefuck

  znap source ohmyzsh/ohmyzsh plugins/colored-man-pages
  znap source ohmyzsh/ohmyzsh plugins/pyenv
  # znap source ohmyzsh/ohmyzsh plugins/virtualenv
  # znap source ohmyzsh/ohmyzsh plugins/virtualenvwrapper

  znap source zsh-users/zsh-completions
  znap source zsh-users/zsh-autosuggestions

  ZSH_HIGHLIGHT_HIGHLIGHTERS=( main brackets )
  znap source zsh-users/zsh-syntax-highlighting

  znap source zsh-users/zsh-history-substring-search
  znap source djui/alias-tips
  znap source lukechilds/zsh-nvm
  znap source jhipster/jhipster-oh-my-zsh-plugin

  znap eval brew-shellenv 'brew shellenv'
  znap eval pyenv-init ${${:-=pyenv}:A}' init -'
  znap eval pip-completion 'pip completion --zsh'
  znap eval pipx-completion 'register-python-argcomplete pipx'

  znap compdef _kubectl 'kubectl completion zsh'
  znap compdef _rustup  'rustup completions zsh'
  znap compdef _cargo   'rustup completions zsh cargo'
```

It Make problems when I put this line 
```
  znap source ohmyzsh/ohmyzsh plugins/fd
  znap source ohmyzsh/ohmyzsh plugins/ripgrep
```

Error Message
```
_arguments:comparguments:325: can only be called from completion function
_arguments:comparguments:325: can only be called from completion function
```

It doesn't make problem when I use OhMyZsh directly or using antigen

**Comment by marlonrichert** [comment, marked as answer]
You can now install completion functions from the command line with `znap install`. For example, to install _all_ completion functions that ship with Oh-My-Zsh, just run the following:
```zsh
znap install ohmyzsh/ohmyzsh
```

<details><summary>Old Answer (Obsolete)</summary>

You shouldn't `source` the docker, fd and ripgrep "plugins", because they are not actual plugins, just completion functions. You should add them to your `$fpath`, like this:
```zsh
fpath+=( ~[ohmyzsh]/{docker,fd,ripgrep} )
```

Furthermore, if you do `znap source ohmyzsh/ohmyzsh` then you are sourcing _all_ of Oh-My-Zsh's libraries (because that's what the `ohmyzsh/oh-my-zsh.sh` scripts does). If you want to source only `lib/completion`, then you should _not_ include the line above.

Finally, you don't need to put all those `znap source ohmyzsh/ohmyzsh XXX` statements on separate lines. You can just do this:
```zsh
znap source ohmyzsh/ohmyzsh lib/completion \
  plugins/{git,command-not-found,fasd,fzf,osx,tmux,tmuxinator,thefuck,colored-man-pages,pyenv}
fpath+=( ~[ohmyzsh]/{docker,fd,ripgrep} )
```
</details>

**Reply by marlonrichert** [reply]
@hoyacom FYI: I just pushed in a new feature. Please see my updated answer, above.

**Reply by chiefy** [reply]
How do you only install a subset of ohmyzsh though? For ex. I just want the docker and docker-compose plugins? Or is the idea you just install/clone the whole repo one-time and then source what you need?

**Reply by marlonrichert** [reply]
Yes, the latter is the idea, because that’s the most reliable way to keep the files updated and to make sure you have all the dependencies that the plugin needs.

However, here’s an example of [how to use `znap eval` to add a single file](https://github.com/marlonrichert/zsh-snap/blob/cbe9ab3c1ce3721fed609d8c522fd070a746f8e1/.zshrc#L92).
